[MRG+1] Sticky margins

anntzer opened the pull request:
Let artists register lists of "sticky" x and y values.  If an edge of
the union of bounding boxes ends up falling on a sticky value, do not
add margins to that edge.

Docs still missing.

See https://github.com/matplotlib/matplotlib/pull/7471#issuecomment-261149863.

I may not have so much time to work on this in the coming days, feel free to directly push to my repo to update this PR.

TODO list

- [ ] allow manual disabling of sticky edges
- [ ] document how to get old style style back.

efiring commented on lib/matplotlib/axes/_axes.py:
```diff
@@ -2128,11 +2121,13 @@ def make_iterable(x):
+            x_stickies, y_stickies = r.get_stickies()
+            (getattr(r.get_stickies(),
+                     {'vertical': 'y', 'horizontal': 'x'}[orientation])
+             .append(0))
```
It looks like you have a 4-physical-line obfuscation of the following 5 lines, correct?
```python
x_stickies, y_stickies = r.get_stickies()
if orientation == 'vertical':
    y_stickies.append(0)
else:
    x_stickies.append(0)
```
But you only need x or y, so you could also do
```python
if orientation == `vertical`:
    r.get_stickies().y.append(0)
else:
    r.get_stickies().x.append(0)
```
so we are back down to 4 physical lines that I find much easier to understand.  And it would look even nicer with `stickies` as a property.

efiring commented on lib/matplotlib/artist.py:
```diff
@@ -926,98 +928,13 @@ def set_zorder(self, level):
         self.stale = True
 
-    def get_top_margin(self):
+    def get_stickies(self):
```
I suggest making `stickies` a read-only property.

efiring commented on lib/matplotlib/axes/_axes.py:
```diff
@@ -6403,8 +6400,11 @@ def _normalize_input(inp, ename='input'):
+            for patch in patches:
+                (getattr(patch.get_stickies(),
+                         {'vertical': 'y', 'horizontal': 'x'}[orientation])
+                 .append(0))
```
I suggest de-obfuscation as in the previous example.

anntzer commented:
@efiring I have addressed all your comments (which I agree with, in particular making stickies a readonly property.

The only artist which I haven't handled yet is contour (and thus the related tests should fail).  I am looking for a way to retrieve the extents of the contour, other than reusing the logic from `QuadContourSet._initialize_x_y` (I *could* store the variables computed at the end of `_process_args` for later use (`self.collections` isn't populated, or even created, yet, at that point), but perhaps I'm missing something less awkward?).  Any ideas?

efiring commented:
This is a good example of why the ContourSet probably should be a compound artist, added to the Axes as a unit, instead of a bunch of collections that are independently added to the Axes.

Given that now is not the time for that major refactoring, I think the thing to do is to save the data limits where they are calculated at the end of each implementation of `_process_args`, use them to set the stickies for each collection where it is added to the Axes, and move the setting of the data limits and the call to `autoscale_view(tight=True)` to the end of the `__init__` methods.

efiring commented:
`numpy.isclose` is new in 1.7, and we still require 1.6.  #7396.  Sure wish we could just do it now.

QuLogic commented:
`np.allclose` is probably a good enough replacement.

efiring commented:
No, it won't work in this case.  The condition is `np.any(np.isclose(a, b))`.

anntzer commented:
It looks like the tests are passing now, with the exception of numpy 1.6 (due to isclose) and docs (which I haven't touched at all).

(As a side note, perhaps we should think briefly whether it should be `isclose(x0, stickies)` or `isclose(stickies, x0)` (check the docstring to see the difference).)

Realistically, I should probably be working on other stuff right now, so I would appreciate if someone could pick up the PR from there.  If not I should be able to get back to it by the end of the month.

efiring commented:
Thanks for bringing it this far in a very short time!  I expect that between @tacaswell and myself (and anyone else who is interested) we can pick it up, if Tom thinks this is the way to go, and we don't find some blocking problem with it.

efiring commented:
Closing and reopening to restart the tests after merge of #7483.

tacaswell commented:
I am a bit confused why these tests are only failing with pytest and not with nose.

tacaswell commented:
I get legit failures locally now (not unbound local).

anntzer commented:
The failure comes from a faulty calculation of `interval.dataLim` in the stacked stepfilled case of hist (line 6441 before this PR, "self.dataLim.intervaly = (ymin, ymax)").  This faulty calculation was present even as of 1.5.3 and can be seen e.g. by calling `autoscale` after a stacked stepfilled plot:
```
hist(np.random.rand(100, 2), stacked=True, histtype="stepfilled"); autoscale()
```
(which basically only autoscales based on the upper stacked histogram).
(You can also print the value of `gca().dataLim` at that point.)

This issue was not visible before because the previous implementation did *not* rely on autoscaling at that point (I think it relied on the calls to `autoscale` during the creation of the individual artists that composed the stacked histogram).  Now that we call `autoscale_view` at the end of `hist()`, the bug gets exposed.

QuLogic commented:
Is that perhaps related to #4414?

anntzer commented:
Likely yes.

anntzer commented on lib/matplotlib/axes/_axes.py:
```diff
@@ -6407,34 +6407,11 @@ def _normalize_input(inp, ename='input'):
-                xmin = max(xmin*0.9, minimum) if not input_empty else minimum
-                xmin = min(xmin0, xmin)
-                self.dataLim.intervalx = (xmin, xmax)
+                self.dataLim.intervalx = (0, xmax)
```
This would fail with negative weights, or if there are already other artists present with greater extents.

efiring replied:
The original code was broken, even before 2.x; the present code is broken.  It turns out that the present "sticky" algorithm is not doing the right thing for plain bar plots, either, when switching to a log scale; this was working in 1.5.3.  I don't know if it was working with the "margins" algorithm in 2.x.

efiring replied:
Yes, everything can be made to work, at least under reasonable circumstances... but the devil is in the details.  I am making progress but running into problems with the test framework.  Rebasing on current master helps, I think.  Do you mind if I force-push the rebase?

anntzer replied:
Not at all, go ahead.

efiring commented:
Warning: I rebased and force-pushed.
We have some awkward code in bar and hist to handle zero bottom values in "log" mode.  We wouldn't actually need this with the new margin code, because the lower-level drawing code uses minpos (minimum positive value) to avoid zero on log scales.  I think we should switch to relying on that mechanism, and discourage the use of the "log" kwarg.  It yields a result that is better when the view limits are changed (vertical bars always go to the bottom with a log scale), and it makes switching back to a linear scale give the right result (bars still go to zero at the bottom).

efiring commented:
I think this might be at least functionally ready.

I still don't like the "stickies" name at all, but I haven't settled on a good replacement.  The problem with it is that I think it will be mystifying to anyone new to the codebase, or even to anyone who doesn't work with this part of it all the time.  It is clever and apt once you know what it is, but not helpful otherwise.  Alternatives?

margin_stops
hard_margins
hard_edges
sticky_edges
edge_lists

Does anyone like any of these, or have better suggestions?

anntzer commented:
I think "sticky_edges" is good.  The problem of names with "hard" or "stop" is that it's not necessarily immediately clear whether they constrain the limit to be on one side or the other of the value.

efiring commented:
OK, I am switching to "sticky_edges".  I will finish this evening, or tomorrow morning.  It looks like there will be another minor change or two to make along the way.

efiring commented:
Again, almost there, but still working on a contour problem.

anntzer commented on lib/matplotlib/axes/_axes.py:
```diff
@@ -5637,8 +5635,10 @@ def pcolormesh(self, *args, **kwargs):
         miny = np.amin(Y)
         maxy = np.amax(Y)
 
-        corners = (minx, miny), (maxx, maxy)
```
Same comment as above (keep definition of {min,max}{x,y} together with corner); not repeating this comment if it still occurs below.

efiring replied:
done

anntzer commented on doc/users/dflt_style_changes.rst:
```diff
@@ -1022,37 +1022,7 @@ The size of the padding in the x and y directions is controlled by the
-to any plotting method or artist ``__init__`` which supports ``**kwargs`` (as
-any unused kwargs eventually get passed to `~matplotlib.artist.Artist.update`).
-
+the view limits coincide with ticks.
```
Should we document this parameter?  This should at least be done at the property level.

efiring replied:
I don't see any reason to add anything in the paragraph you are commenting on.  I did add to the property docstring, which is already referenced in the api documentation.

anntzer replied:
The docstring is good enough, I only commented at this line because that was the most practical to do.

anntzer replied:
Perhaps the docs for `margins` should also reference this? So that someone who tries to explicitly set margins for an image and sees no margins being applied can find out why this is the case.

efiring commented on lib/matplotlib/axes/_axes.py:
```diff
@@ -5486,8 +5483,10 @@ def pcolor(self, *args, **kwargs):
         miny = np.amin(y)
         maxy = np.amax(y)
```
done

efiring commented:
Something non-deterministic is still bedeviling the tests.  3 of the 5 appveyor tests failed on single tests with extremely small deviations.  There is no point in restarting the whole suite.  We need to either allow a little tolerance, or figure out how to make the tests completely deterministic--like computer calculations *should* be.
The only thing I have thought of as a vague possible cause is alignment.  Intel architecture is not strict in its requirement for number alignment in memory.  I'm wondering whether there might be some library call that rounds differently, for example, depending on how an 8-byte number is aligned in memory.  If so, we are stuck with it.

tacaswell commented on lib/matplotlib/axes/_base.py:
```diff
@@ -2199,53 +2199,18 @@ def autoscale_view(self, tight=None, scalex=True, scaley=True):
-                'right': False
-            }
+            stickies = [artist.stickies for artist in self.get_children()]
+            x_stickies = sum([sticky.x for sticky in stickies], [])
```
is this better written as 

```python
x_stickies = [x for sticky in stickies for x in sticky.x]
```
?

tacaswell replied:
This is :bike: :house: ing, treat as such

anntzer replied:
Probably best to use np.concatenate.

tacaswell commented:
I am :+1: on this.  

Documentation of the non-classic mode recoverable change to tricontour plot (seen the the two test image changes).

The change to `pre_transform` is reverting to what the test was in 1.5 so that is a good thing.

Happy to see the docs as a follow on PR.

NelleV commented:
I wouldn't merge this without proper documentation. We have way too many undocumented features in matplotlib, which leads to dead code that no one maintains (though in this particular case, I am not too worried about this code being dead in a couple of years).
This PR has been opened only 12 days, so I am sure @anntzer is fine writing a bit of documentation for it, or would be opened to people writing documentation for him and opening a PR on his fork.

anntzer commented:
I am currently travelling so won't be able to work on it until say the end of the week.  If someone else wants to work on the docs, please go ahead, otherwise I'll do it in a week or so.

efiring commented:
Now there is still one appveyor test failure (only in the 2.7 "test all" run):
```
ImageComparisonFailure: images not close: C:\projects\matplotlib\result_images\test_png\pngsuite.png vs. C:\projects\matplotlib\result_images\test_png\pngsuite-expected.png (RMS 0.011)
```

tacaswell commented:
That one _also_ needs to have it's thershold bumped, but is also failing on master (unrelated to this).

efiring commented:
After yesterday's conversation, I think that we need to add one more component: an API for switching the `sticky_edges` processing on or off in the autoscaling.  The reason is that the `sticky_edges` machinery is an attempt to provide behavior that matches the type of artist being plotted, and to do a reasonable job of handling a mixture of artists.  There will inevitably be cases where the machinery doesn't automatically do what the user wants, however, so there needs to be an easy way of saying "ignore the sticky_edges and use the specified margins".  At present the user could do this by emptying the `sticky_edges` lists prior to a final call to `autoscale_view`, but this is too fiddly, and it involves erasing information that probably should be preserved, so that the operation is reversible.  Instead, we probably need something like an `Axes._use_sticky_edges` flag, with a simple API for manipulating it.
I can add that some time within the next few days.

anntzer commented:
See also https://github.com/matplotlib/matplotlib/pull/6281#issuecomment-215617480 re: reversibility.

NelleV commented on lib/matplotlib/axes/_base.py:
```diff
@@ -2199,53 +2199,22 @@ def autoscale_view(self, tight=None, scalex=True, scaley=True):
+            stickies = [artist.sticky_edges for artist in self.get_children()]
+            x_stickies = sum([sticky.x for sticky in stickies], [])
+            y_stickies = sum([sticky.y for sticky in stickies], [])
+            if self.get_xscale().lower() == 'log':
```
Shouldn't the presence of negative stickies with log scale raise an error?

anntzer replied:
Not any more than negative values in a plot do, I think?

NelleV replied:
Well, I think it is a bug that we tolerate this :)

efiring replied:
@nelleV, you think that if a histogram plot is switched to a log scale it should fail rather than yielding a plot?  In other words, simply bumping the "L" key when a histogram plot is active should raise an exception?

NelleV replied:
If it means that the following code raises an error, then yes I think it should.
I think the fact that matplotlib accepts log scale when the user provides negative value is at high risk of having unintended bugs, and leads to wrong results and interpretation of analysis.
On the other hand, I'd be fine if the error was caught internally when in interactive mode.

```python
import matplotlib.pyplot as plt
import numpy as np

fig, ax = plt.subplots()
ax.set_yscale("log")
ax.plot(np.arange(0, 10))
ax.plot(np.arange(-100, 0))
```

anntzer replied:
I think the behavior proposed by @NelleV is reasonable but this should not be handled by the sticky system (nor block this PR) -- it should affect all plots, including those who do not set sticky margins.

NelleV replied:
I agree.

tacaswell commented on lib/matplotlib/axes/_base.py:
```diff
@@ -2034,6 +2035,28 @@ def set_autoscaley_on(self, b):
+
+    @use_sticky_edges.setter
+    def use_sticky_edges(self, b):
+        self._use_sticky_edges = bool(b)
```
Should this also re-trigger an auto-scaling?

efiring replied:
I was wondering about that also, but decided against it.  I could be convinced otherwise for now, but long term I think we might want to consolidate the autoscaling at draw time instead of having calls to `autoscale_view` sprinkled all over the place.  I don't think it makes sense to trigger autoscaling with every change in the state of the plot.

tacaswell replied:
I am convinced

efiring replied:
Good point.  You are welcome to replace it with a comment to that effect.  I think I put the call in because we tend to have it in every function that potentially changes how the plot will look.

anntzer replied:
Done, together with indentation fix below.

anntzer replied:
Good point.
On the other hand this is easy enough to revert when we do switch to draw-time autoscaling (which makes complete sense).

efiring commented:
I've bumped this back to "MRG+1" based on Tom's earlier approval and all tests passing.  If people are happy enough with the API, including the `Axes.use_sticky_edges` property that I added, then I recommend merging *without* waiting for all documentation additions and fixups.  We need to get people started using it so we can see whether it will require any more changes prior to v2.0.

QuLogic commented on lib/matplotlib/axes/_axes.py:
```diff
@@ -5800,34 +5799,33 @@ def pcolorfast(self, *args, **kwargs):
-                                    **kwargs)
+            if style == "image":
+                im = mimage.AxesImage(self, cmap, norm,
+                                            interpolation='nearest',
```
I guess it passes, but it's a bit of a weird indent.

efiring replied:
True.  It's accidental.  You are welcome to fix it.

tacaswell commented on lib/matplotlib/tri/tricontour.py:
```diff
@@ -50,11 +50,8 @@ def _process_args(self, *args, **kwargs):
-            x1 = tri.x.max()
-            y0 = tri.y.min()
-            y1 = tri.y.max()
-            self.ax.update_datalim([(x0, y0), (x1, y1)])
```
I do not think we need the line below this either, the autoscaling gets called later in the init.

tacaswell commented:
Assuming my most recent commit passes, I agree with @efiring that this should be merged.

Given that @anntzer and @efiring both have large numbers of commit is this, can @NelleV or @QuLogic  give this a final review and push the green button?

NelleV commented:
Green button pushed.

NelleV commented:
On the other hand, related tickets have probably not been closed. Can someone elte take care of this?

tacaswell commented:
I am in the process of backporting this

tacaswell commented:
backported to v2.x as 0289b4f
